### PR TITLE
camera: support multiple cameras within one instance

### PR DESCRIPTION
**Pull request by julianoes**
This changes the API to support more than one camera within one camera plugin instance.

This will enable multiple cameras in language wrappers instead of just C++ as it is now.

Contains several changes:
- Support MavlinkFTP for camera definition files.
- Disable photo capture list for now (would need server support and system test testing).
- Move camera integration tests to system tests.
- Remove leftover logging integration test.
- Remove camera mavsdk_server tests because it was too hard too adapt it all.

Contains:
https://github.com/mavlink/MAVSDK-Proto/pull/352
https://github.com/mavlink/MAVSDK-Proto/pull/358

### CODE REVIEW

**Comment by julianoes** [comment]
Closes #2365

**Comment by JonasVautherin** [comment]
My guess is that `HTTP_ONLY=OFF` in this PR makes curl link libnghttp2, and it fails:

```
-- Checking for module 'libnghttp2'
--   Found libnghttp2, version 1.64.0
-- Found NGHTTP2: /opt/homebrew/Cellar/libnghttp2/1.64.0/include (found version "1.64.0")
```

Interestingly libnghttp2 is not available on other platforms and that's fine. I wonder why it gets pulled on iOS (it does not seem to be pulled in macOS 🤔)

File: examples/start_stop_server/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+project(start_stop_server)
+
+add_executable(start_stop_server
+    start_stop_server.cpp
+)
+
+find_package(MAVSDK REQUIRED)
+
+target_link_libraries(start_stop_server
+    -L/home/julianoes/src/upstream/MAVSDK/install/lib/libmavsdk.so.2
```
[the remark this answers is not in the document]
Comment (julianoes) [reply examples/start_stop_server/CMakeLists.txt:15]: This was my workaround related to https://github.com/mavlink/MAVSDK/pull/2441.

File: third_party/CMakeLists.txt [hunk 2]
```diff
@@ -35,6 +35,13 @@ if(SUPERBUILD)
         build_target(grpc)
     endif()
 
+    if(NOT BUILD_WITHOUT_CURL)
+        if(NOT IOS)
+            build_target(zlib-ng)
+        endif()
+        build_target(curl)
+    endif()
+
```
[the remark this answers is not in the document]
Comment (julianoes) [reply third_party/CMakeLists.txt:44]: Thanks, I'm removing it.

**Comment by julianoes** [comment]
@JonasVautherin I think this is ready to go in, and then we can release v3 and I can start using and testing it.

File: src/mavsdk/plugins/camera/camera_impl.cpp [hunk 3]
```diff
@@ -40,636 +43,479 @@ CameraImpl::~CameraImpl()
...
+    _system_impl->register_mavlink_message_handler(
         MAVLINK_MSG_ID_VIDEO_STREAM_STATUS,
-        _camera_id + MAV_COMP_ID_CAMERA,
         [this](const mavlink_message_t& message) { process_video_stream_status(message); },
         this);
 
-    _check_connection_status_call_every_cookie =
-        _system_impl->add_call_every([this]() { check_connection_status(); }, 0.5);
+    //_request_missing_capture_info_cookie =
+    //    _system_impl->add_call_every([this]() { request_missing_capture_info(); }, 0.5);
```
Comment (JonasVautherin) [inline src/mavsdk/plugins/camera/camera_impl.cpp:110]: Should that come back or go away?

File: src/mavsdk/plugins/camera/camera_impl.cpp [hunk 4]
```diff
@@ -40,636 +43,479 @@ CameraImpl::~CameraImpl()
...
-        MAVLINK_MSG_ID_CAMERA_SETTINGS, cmp_id, this);
-
-    _system_impl->update_component_id_messages_handler(
-        MAVLINK_MSG_ID_CAMERA_INFORMATION, cmp_id, this);
+    _system_impl->cancel_all_param(this);
 
-    _system_impl->update_component_id_messages_handler(
-        MAVLINK_MSG_ID_VIDEO_STREAM_INFORMATION, cmp_id, this);
+    std::lock_guard lock(_mutex);
+    //_system_impl->remove_call_every(_request_missing_capture_info_cookie);
```
Comment (JonasVautherin) [inline src/mavsdk/plugins/camera/camera_impl.cpp:126]: I guess that goes with the comment above

File: src/mavsdk/plugins/camera/camera_impl.cpp [hunk 5]
```diff
@@ -1083,200 +902,204 @@ float CameraImpl::to_mavlink_camera_mode(const Camera::Mode mode) const
...
-            _status.image_count_at_connection = camera_capture_status.image_count;
-        }
-    }
+    //    _status.image_count = camera_capture_status.image_count;
 
-    check_status();
+    //    if (_status.image_count_at_connection == -1) {
+    //        _status.image_count_at_connection = camera_capture_status.image_count;
+    //    }
+    //}
```
Comment (JonasVautherin) [inline src/mavsdk/plugins/camera/camera_impl.cpp:1026]: What about those?

File: src/mavsdk/plugins/camera/camera_impl.cpp [hunk 6]
```diff
@@ -1285,61 +1108,85 @@ void CameraImpl::process_camera_image_captured(const mavlink_message_t& message)
...
+        _capture_info.last_advertised_image_index = capture_info.index;
     }
+
+    else if (auto it = _capture_info.missing_image_retries.find(capture_info.index);
+             it != _capture_info.missing_image_retries.end()) {
+        _capture_info.callbacks.queue(
+            capture_info, [this](const auto& func) { _system_impl->call_user_callback(func); });
+        _capture_info.missing_image_retries.erase(it);
+    }
+#endif
```
Comment (JonasVautherin) [inline src/mavsdk/plugins/camera/camera_impl.cpp:1183]: Shouldn't this come back?

File: src/mavsdk/plugins/camera/camera_impl.cpp [hunk 7]
```diff
@@ -1285,61 +1108,85 @@ void CameraImpl::process_camera_image_captured(const mavlink_message_t& message)
...
 }
 
 void CameraImpl::request_missing_capture_info()
 {
-    std::lock_guard<std::mutex> lock(_capture_info.mutex);
+#if 0
```
Comment (JonasVautherin) [inline src/mavsdk/plugins/camera/camera_impl.cpp:1188]: Not sure what to think about this

File: src/mavsdk/plugins/camera/camera_impl.cpp [hunk 8]
```diff
@@ -1350,52 +1197,38 @@ void CameraImpl::request_missing_capture_info()
             ++it;
         }
     }
+#endif
 
+    /* TODO: implement again
```
Comment (JonasVautherin) [inline src/mavsdk/plugins/camera/camera_impl.cpp:1202]: And this

File: src/mavsdk/plugins/camera/camera_impl.cpp [hunk 9]
```diff
@@ -1417,274 +1250,320 @@ void CameraImpl::process_camera_information(const mavlink_message_t& message)
...
-    _information.data.focal_length_mm = camera_information.focal_length;
-    _information.data.horizontal_sensor_size_mm = camera_information.sensor_size_h;
-    _information.data.vertical_sensor_size_mm = camera_information.sensor_size_v;
-    _information.data.horizontal_resolution_px = camera_information.resolution_h;
-    _information.data.vertical_resolution_px = camera_information.resolution_v;
-
-    _information.subscription_callbacks.queue(
-        _information.data, [this](const auto& func) { _system_impl->call_user_callback(func); });
+    Camera::Information new_information{};
+    // TODO: Check the case for 1-6.
```
Comment (JonasVautherin) [inline src/mavsdk/plugins/camera/camera_impl.cpp:1254]: What does that mean?

File: src/mavsdk/plugins/camera/camera_impl.cpp [hunk 10]
```diff
@@ -2300,45 +2295,63 @@ void CameraImpl::reset_settings_async(const Camera::ResultCallback callback)
...
-        std::lock_guard<std::mutex> lock(_capture_info.mutex);
-        _capture_info.last_advertised_image_index = -1;
-        _capture_info.missing_image_retries.clear();
-    }
+    UNUSED(camera);
+    // camera.capture_status.photo_list.clear();
+    // camera.capture_status.image_count = 0;
+    // camera.capture_status.image_count_at_connection = 0;
+    // camera.capture_info.last_advertised_image_index = -1;
+    // camera.capture_info.missing_image_retries.clear();
```
Comment (JonasVautherin) [inline src/mavsdk/plugins/camera/camera_impl.cpp:2305]: Feels like this should come back too 🙈

File: src/mavsdk/plugins/camera/camera_impl.cpp [hunk 11]
```diff
@@ -2300,45 +2295,63 @@ void CameraImpl::reset_settings_async(const Camera::ResultCallback callback)
...
         return;
     }
 
+    // TODO: enable again with a camera to test against.
+    _system_impl->call_user_callback([callback]() {
+        LogErr() << "Photo list not implemented.";
+        callback(Camera::Result::ProtocolUnsupported, std::vector<Camera::CaptureInfo>{});
+    });
+
+#if 0
```
Comment (JonasVautherin) [inline src/mavsdk/plugins/camera/camera_impl.cpp:2343]: 😕

File: src/mavsdk/plugins/camera/camera_impl.h [hunk 12]
```diff
@@ -23,299 +28,369 @@ class CameraImpl : public PluginImplBase {
...
+    //    int image_count{-1};
+    //    int image_count_at_connection{-1};
+    //    std::map<int, Camera::CaptureInfo> photo_list;
+    //    bool is_fetching_photos{false};
+    // };
+
+    // struct CaptureInfo {
+    //     int last_advertised_image_index{-1};
+    //     std::map<int, int> missing_image_retries{};
+    // };
```
Comment (JonasVautherin) [inline src/mavsdk/plugins/camera/camera_impl.h:180]: Is there another kind of way to tell the user that a photo has been taken or is it completely disabled?

**Review by JonasVautherin — COMMENTED** [review]
I am a little bit concerned because it seems like quite a bunch of code in the camera plugin has been disabled 🤔.

Is that expected?

File: src/mavsdk/plugins/camera/camera_impl.h (hunk 12 above)
Comment (julianoes) [reply src/mavsdk/plugins/camera/camera_impl.h:180]: I want to add it back in, but I have nothing to test this with. I was hoping I could write the server part of it, and then test it against it once I have that.

It's not really something that is very common or standardized in MAVLink, right?

File: src/mavsdk/plugins/camera/camera_impl.h [hunk 13]
```diff
@@ -23,299 +28,369 @@ class CameraImpl : public PluginImplBase {
...
-    void request_camera_settings();
-    void request_camera_information();
-    void request_video_stream_info();
-    void request_video_stream_status();
-    void request_status();
+    // void request_camera_settings(int32_t camera_id);
+    void request_camera_information(uint8_t component_id);
+    // void request_video_stream_info(int32_t camera_id);
+    // void request_video_stream_status(int32_t camera_id);
+    // void request_status(int32_t camera_id);
```
[the remark this answers is not in the document]
Comment (julianoes) [reply src/mavsdk/plugins/camera/camera_impl.h:316]: Yep a bit messy still.

File: src/mavsdk/plugins/camera/camera_impl.cpp (hunk 9 above)
Comment (julianoes) [reply src/mavsdk/plugins/camera/camera_impl.cpp:1254]: Oh thanks, yes I should fix that. 1-6 are magic entries for cases where the autopilot is implementing the camera manager itself.

File: src/mavsdk/plugins/camera/camera_impl.cpp (hunk 7 above)
Comment (julianoes) [reply src/mavsdk/plugins/camera/camera_impl.cpp:1188]: Neither :smile:

File: src/mavsdk/plugins/camera/camera_impl.h (hunk 12 above)
Comment (JonasVautherin) [reply src/mavsdk/plugins/camera/camera_impl.h:180]: I can't speak for "standardized" 🙈, but I think it's common for mission planning apps to expect some feedback from the camera, e.g. to know that photos are being taken more or less as expected. The updates are unreliable so some updates are missed, but the apps would generally rather have a reliable update than no update at all, I would say.

Comment (julianoes) [reply src/mavsdk/plugins/camera/camera_impl.h:180]: Sure, I agree there.

**Comment by JonasVautherin** [comment]
You mean disabling nghttp2? I think we can. If someone ever needs HTTP2, they can try to enable it later...

**Comment by julianoes** [comment]
I was just confused why main doesn't need the changes I have here.

**Pull request merged**